app/tools/filedb.js

```diff
@@ -1,64 +1,209 @@
+   * @returns {Promise} promise to write a compressed file with a File as resolve parameter
+   */
+  static storeFile(pFile) {
+    if (!(pFile instanceof File)) {
+      return Promise.reject(new TypeError('Provided file is not an instance of FileSchema.'));
+    }
+
+    if (!pFile.checksum()) {
+      return Promise.reject(new Error('Could not resolve a checksum for the file.'));
+    }
+
+    logger.info(`Storing file ${pFile.name} (filename: ${pFile.checksum()}.${fileEnding}).`);
+    return FileDB._checkFilenameAvailability(pFile.checksum()).then((available) => {
+      if (available) {
+        return FileDB._compress(pFile.data).then(compressedData => FileDB._writeFile(pFile.checksum(), compressedData));
+      }
+
+      logger.warn('File already exists, no need to store a duplicate.');
+      return Promise.resolve;
     });
-};
+  }
 
-module.exports.storeFile = function storeFile(file, callback) {
+  /**
+   * Ensures that there is no file stored with the provided filename
+   * @param {string|Buffer|integer} pFilename - filename or file descriptor, is usually the sha1 checksum of the data
+   * @returns {Promise} promise that there is no file with the given name with
+   *                    boolean variable as the resolve parameter
+   */
+  static _checkFilenameAvailability(pFilename) {
+    const filePath = FileDB._resolveFilename(pFilename);
 
-  if(!_.has(file, 'size')) {
-      file.size = file.data.length;
+    logger.debug(`Checking if file exists in path: ${filePath}.`);
+    return fs.exists(filePath).then((exists) => {
+      if (exists) logger.debug(`File ${pFilename} already exists in path: ${filePath}.`);
+      else        logger.debug(`Path: ${filePath} is confirmed to be empty.`); // eslint-disable-line
+
+      return !exists;
+    });
   }
-  if(!_.has(file, 'sha1')) {
-      file.sha1 = checksum(file.data, 'sha1');
+
+  /**
+   * Reads a stored file with a specific filename
+   * @param {string|Buffer|integer} pFilename - filename or file descriptor, is usually the sha1 checksum of the data
+   * @returns {Promise} promise to read a file with a dataBuffer as the resolve parameter
+   */
+  static _readFile(pFilename) {
+    const filePath = FileDB._resolveFilename(pFilename);
+
+    logger.debug(`Reading file from: ${filePath}.`);
+    return fs.readFile(filePath).then((dataBuffer) => {
+      logger.debug(`Read file (filename: ${pFilename} size: ${dataBuffer.size}).`);
+      return dataBuffer;
+    }).catch((error) => {
+      logger.warn(`Could not read file with path: ${filePath}, error: ${error.message}.`);
+      throw error;
+    });
   }
-  if(!_.isFunction(callback)) {
-      callback = function(){};
+
+  /**
+   * Writes an instance of File to filedb
+   * @param {File} pFilename - name of the file to write, without the file ending
+   * @param {string|Buffer} pData - data to write
+   * @returns {Promise} promise to write the file to filedb
+   */
+  static _writeFile(pFilename, pData) {
```

Comment:
we probably need also stream support for writing phase. add @todo section here..